openfisca_core/simulations/_type_guards.py

```diff
@@ -0,0 +1,22 @@
+from __future__ import annotations
+
+from typing import Iterable
+from typing_extensions import TypeGuard
+
+from .typing import AbbrParams, AxesParams, ExplParams, ImplParams, Params
+
+
+def is_abbr_spec(params: Params, items: Iterable[str]) -> TypeGuard[AbbrParams]:
+    return any(key in items for key in params.keys()) or params == {}
+
+
+def is_impl_spec(params: Params, items: Iterable[str]) -> TypeGuard[ImplParams]:
```

Comment:
I don't really understand what this function does - what seems awkward about it is that it's named like a predicate, but returns a set operation.

Reply:
I'll document these functions before merging. TL; DR : [type narrowing](https://mypy.readthedocs.io/en/stable/type_narrowing.html)

Reply:
I'll document these functions before merging. TL; DR : [type narrowing](https://mypy.readthedocs.io/en/stable/type_narrowing.html)

We're asking: is this dict of that type? If the answer is yes, we narrow the type of the dict. In this particular case is very useful because it allows to "bring to life" a full range of concepts that lie in our heads: fully defined entities, explicit and implicit, etc. At runtime, these act just like a regular _isinstance_ would if these were class instances instead of plain dicts.